### PR TITLE
[MM-23281] Persist registeredPlugin in plugin.Environment when plugin is deactivated

### PR DESCRIPTION
**Pull request by mickmister**
#### Summary

As discussed in https://github.com/mattermost/mattermost-server/pull/14088, we are moving the `failedTimestamps` to the plugin health check job to remove the coupling between the job and `plugin.Environment`.

When reviewing this change, @ali-farooq0 and I went with the direction of persisting the `registeredPlugin`, so that `PluginStateFailedToStayRunning` can be present in `plugin.Environment` after the crashed plugin is deactivated. This lets the `Environment`'s state be the source of truth.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-23281
https://mattermost.atlassian.net/browse/MM-23365

### CODE REVIEW

File: plugin/environment.go [hunk 1]
```diff
@@ -238,6 +236,9 @@ func (env *Environment) Activate(id string) (manifest *model.Manifest, activated
 	// Store latest BundleInfo in case something has changed since last activation
 	rp.BundleInfo = pluginInfo
 	env.registeredPlugins.Store(id, rp)
+	if env.pluginHealthCheckJob != nil {
+		env.pluginHealthCheckJob.EnsurePlugin(id)
```
Comment (ali-farooq0) [inline plugin/environment.go]: do we need to register plugins to the healthchecker? I wonder if we can rely on querying `job.env.Active()` on every heartbeat and store the failed plugins inside the healthchecker? 🤔

**Comment by mickmister** [comment]
From the Jira ticket
> - `Environment.SetPluginState` should be a private func as no one externally should be allowed to set plugin states
> - Webapp clients should be notified (through websocket) when `PluginHealthCheckJob` sets plugin's error state 

Does this mean that the health check job should not have access to `SetPluginState`? That is currently how the job sets the plugin's state to `model.PluginStateFailedToStayRunning`.

https://github.com/mattermost/mattermost-server/blob/07e7a24c9ff7571c147a65a6e37875a7970754e6/plugin/health_check.go#L126-L127

**Comment by mickmister** [comment]
tldr: @ali-farooq0 The changes you propose for testing may be out of scope of the PR. It will require some reorganizing of the health check job in regards to its coupling with `Environment`. I'd like to talk this through when we both have some time next week.

-----

> Add a test that tests healthchecker deactivating a plugin so we can lock down the behaviour.

I tried to do this when I first created the health check job. In order to instantiate the job, we need a `plugin.Environment`. In order to instantiate a `plugin.Environment`, we need an `app.App`. We can't get that in the `plugin` package because there is a circular dependency.

The health check tests are currently in the `plugin` package, though they should probably be in `plugin_test`, not to be confused with `plugintest` where the mocks are stored. Testing from package `plugin_test` means we don't have access to the private methods of `HealthCheckJob`.

Currently in my attempts of testing, I'm making an interface to define what functionality `HealthCheckJob` needs from `Environment`, then testing in the `plugin_test` package. But this makes all of the other health check tests not compile, because they are testing private structs in the `plugin` package. We could have two files of health check tests but that seems off.

I think we need to discuss how we want to do this, because adding this test seems out of scope of this PR.

**Comment by ali-farooq0** [comment]
> From the Jira ticket
> 
> > * `Environment.SetPluginState` should be a private func as no one externally should be allowed to set plugin states
> > * Webapp clients should be notified (through websocket) when `PluginHealthCheckJob` sets plugin's error state
> 
> Does this mean that the health check job should not have access to `SetPluginState`? That is currently how the job sets the plugin's state to `model.PluginStateFailedToStayRunning`.
> 
> https://github.com/mattermost/mattermost-server/blob/07e7a24c9ff7571c147a65a6e37875a7970754e6/plugin/health_check.go#L126-L127

It should, but it should be the only one, since the healthchecker is in the same package it should be able to access Environment's internal functions.

**Comment by mickmister** [comment]
> It should, but it should be the only one, since the healthchecker is in the same package it should be able to access Environment's internal functions.

Agreed, updated.

**Comment by ali-farooq0** [comment]
> I think we need to discuss how we want to do this, because adding this test seems out of scope of this PR.

Sounds good, will schedule something for next week, to discuss in detail. Thanks again for the last minute push on this pr! much appreciated :)

File: plugin/health_check.go [hunk 2]
```diff
@@ -137,10 +147,10 @@ func (job *PluginHealthCheckJob) Cancel() {
...
-	if len(rp.failTimeStamps) >= HEALTH_CHECK_RESTART_LIMIT {
-		index := len(rp.failTimeStamps) - HEALTH_CHECK_RESTART_LIMIT
-		t := rp.failTimeStamps[index]
+func shouldDeactivatePlugin(failedTimestamps []time.Time) bool {
+	if len(failedTimestamps) >= HEALTH_CHECK_RESTART_LIMIT {
+		index := len(failedTimestamps) - HEALTH_CHECK_RESTART_LIMIT
+		t := failedTimestamps[index]
 		now := time.Now()
 		elapsed := now.Sub(t)
 		if elapsed <= HEALTH_CHECK_DISABLE_DURATION {
```
Comment (ali-farooq0) [inline plugin/health_check.go]: just wondering, why do we have to do this check isn't this always true since the time between now and last failure will always be less than HEALTH_CHECK_DISABLE_DURATION because we ping every 30 seconds?

Can we rely on number of failures rather than the timestamp? would make the logic alot simpler :)

Comment (mickmister) [reply plugin/health_check.go]: > just wondering, why do we have to do this check isn't this always true since the time between now and last failure will always be less than HEALTH_CHECK_DISABLE_DURATION because we ping every 30 seconds?

The reasoning behind the timestamps is that a failure might be happening every 10 minutes or so, maybe because of an endpoint that occasionally gets called is panicking. The consensus was that if the plugin fails three times within one hour, then we disable the plugin. This was the most straightforward implementation I could think of that satisfies the requirement.

File: plugin/health_check.go [hunk 3]
```diff
@@ -73,6 +83,7 @@ func (job *PluginHealthCheckJob) Start() {
 
 // checkPlugin determines the plugin's health status, then handles the error or success case.
 func (job *PluginHealthCheckJob) checkPlugin(id string) {
+	job.EnsurePlugin(id)
 	p, ok := job.env.registeredPlugins.Load(id)
```
[the remark this answers is not in the document]
Comment (mickmister) [reply plugin/health_check.go]: I opted for making `env.performHealthCheck` so the health check job doesn't have to know about `or its methods.

File: plugin/health_check.go [hunk 4]
```diff
@@ -94,23 +105,22 @@ func (job *PluginHealthCheckJob) checkPlugin(id string) {
...
-	job.env.registeredPlugins.Store(id, p)
+	ftime = append(ftime, time.Now())
+	job.failTimestamps.Store(id, ftime)
 
-	if shouldDeactivatePlugin(p) {
-		p.failTimeStamps = []time.Time{}
-		job.env.registeredPlugins.Store(id, p)
+	if shouldDeactivatePlugin(ftime) {
+		ftime = []time.Time{}
+		job.failTimestamps.Store(id, ftime)
```
[the remark this answers is not in the document]
Comment (mickmister) [reply plugin/health_check.go]: Decided not to do this since this structure is now being used to determine if the plugin has crashed.

File: plugin/health_check.go (hunk 2 above)
Comment (ali-farooq0) [reply plugin/health_check.go]: ahhh I see. The implementation make sense now :) Thanks for the clarification.

Maybe worth adding a comment here something to the effect of 
`// Disable plugin if last HEALTH_CHECK_RESTART_LIMIT failures happen within the HEALTH_CHECK_DISABLE_DURATION time limit`

also i'd suggest using `time.Since` which is just a shorthand for `time.Now().Sub(t)`, but makes it a tiny bit easier to read.

https://golang.org/pkg/time/#Since

File: plugin/health_check.go [hunk 5]
```diff
@@ -12,105 +12,92 @@ import (
...
-	}()
+	}
 }
 
-// checkPlugin determines the plugin's health status, then handles the error or success case.
-func (job *PluginHealthCheckJob) checkPlugin(id string) {
-	p, ok := job.env.registeredPlugins.Load(id)
+// CheckPlugin determines the plugin's health status, then handles the error or success case.
+func (job *PluginHealthCheckJob) CheckPlugin(id string) {
+	statusInterface, ok := job.healthCheckStatuses.Load(id)
```
Comment (ali-farooq0) [inline plugin/health_check.go]: if the load check is moved from here to handleHealthCheckFail function then this whole function goes away as it'll only be:
```
	err := job.env.performHealthCheck(id)
	if err != nil {
		mlog.Error("Health check failed for plugin", mlog.String("id", id), mlog.Err(err))
		job.handleHealthCheckFail(id, status)
	}
```
which can be moved inside `Start()` Thoughts?

File: app/plugin_health_check_test.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+package app
+
+import (
+	"testing"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/stretchr/testify/require"
+)
+
+func TestHealthCheckJob(t *testing.T) {
```
Comment (ali-farooq0) [inline app/plugin_health_check_test.go]: Excellent test! 👍

File: app/plugin_health_check_test.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+	bundles := env.Active()
+	require.Equal(t, 1, len(bundles))
+
+	id := bundles[0].Manifest.Id
+	job.CheckPlugin(id)
+	bundles = env.Active()
+	require.Equal(t, 1, len(bundles))
+
+	statuses, err := env.Statuses()
```
Comment (ali-farooq0) [inline app/plugin_health_check_test.go]: nit: why not just query for the plugin state? `env.GetPluginState()`? otherwise you'll have to assert on the len as well.

File: plugin/health_check.go [hunk 8]
```diff
@@ -12,105 +12,92 @@ import (
...
 )
 
-type PluginHealthCheckJob struct {
-	cancel     chan struct{}
-	cancelled  chan struct{}
-	cancelOnce sync.Once
-	env        *Environment
+type PluginHealthCheckStatus struct {
+	failTimestamps []time.Time
+	Crashed        bool
```
Comment (ali-farooq0) [inline plugin/health_check.go]: if the healthchecker ends up setting the state on env, this might become unnecessary to track.

File: plugin/health_check.go [hunk 9]
```diff
@@ -134,16 +121,14 @@ func (job *PluginHealthCheckJob) Cancel() {
...
-		elapsed := now.Sub(t)
-		if elapsed <= HEALTH_CHECK_DISABLE_DURATION {
+// shouldDeactivatePlugin determines if a plugin needs to be deactivated after the plugin has failed (HEALTH_CHECK_RESTART_LIMIT) times,
+// within the configured time window (HEALTH_CHECK_DEACTIVATION_WINDOW).
+func shouldDeactivatePlugin(failedTimestamps []time.Time) bool {
+	if len(failedTimestamps) >= HEALTH_CHECK_RESTART_LIMIT {
+		index := len(failedTimestamps) - HEALTH_CHECK_RESTART_LIMIT
+		t := failedTimestamps[index]
+		elapsed := time.Since(t)
+		if elapsed <= HEALTH_CHECK_DEACTIVATION_WINDOW {
```
Comment (ali-farooq0) [inline plugin/health_check.go]: could simplify:
```suggestion
		if time.Since(failedTimestamps[index]) <= HEALTH_CHECK_DEACTIVATION_WINDOW {
```

File: plugin/environment.go [hunk 10]
```diff
@@ -483,6 +484,21 @@ func (env *Environment) RunMultiPluginHook(hookRunnerFunc func(hooks Hooks) bool
 	}
 }
 
+// performHealthCheck uses the active plugin's supervisor to verify if the plugin has crashed.
+func (env *Environment) performHealthCheck(id string) error {
```
Comment (ali-farooq0) [inline plugin/environment.go]: 👌

**Comment by mickmister** [comment]
@ali-farooq0 Thanks for the thorough review :+1:

Setting the state from the health check job is currently not possible, because the `Environment` removes the `registeredPlugin` when the plugin is deactivated, so there is no structure to store the `PluginStateFailedToStayRunning` value. If we want to continue with the direction of removing the `registeredPlugin`, then there needs to be another place where we store the failed state of the plugin, for example the health check job. This is the reason why I initially had the `registeredPlugin` /not/ being removed when the plugin is deactivated.

I'm not partial to storing it in the health check job. That's just the only place that makes sense if we are removing the `registeredPlugin` on deactivation.

File: plugin/health_check.go (hunk 5 above)
Comment (mickmister) [reply plugin/health_check.go]: This currently won't work with the current state of using the `Crashed` boolean, because the `Crashed` value will not be updated to `false` upon externally re-activating the plugin. The `HealthCheckStatus`'s `Crashed` value must be updated when we know that the plugin is currently in a state where we can check its health. i.e. it has not failed 3 times.

**Comment by ali-farooq0** [comment]
Gotcha. Alright makes sense then 👍

File: plugin/environment.go [hunk 11]
```diff
@@ -134,14 +132,17 @@ func (env *Environment) IsActive(id string) bool {
 func (env *Environment) GetPluginState(id string) int {
 	rp, ok := env.registeredPlugins.Load(id)
 	if !ok {
+		if env.pluginHealthCheckJob != nil && env.pluginHealthCheckJob.hasPluginCrashed(id) {
```
[the remark this answers is not in the document]
Comment (mickmister) [reply plugin/environment.go]: @ali-farooq0 Do you have a suggestion to get around this?

File: plugin/health_check.go (hunk 5 above)
Comment (mickmister) [reply plugin/health_check.go]: @ali-farooq0 ^ I think this is the last main piece of feedback that needs to be resolved. The current state of the code that you point out is not too elegant, but we need to make sure that the healthiness (`Crashed = false`) is reset when the plugin is back online.

File: plugin/health_check.go [hunk 12]
```diff
@@ -12,105 +12,91 @@ import (
...
 	// Append current failure before checking for deactivate vs restart action
-	p.failTimeStamps = append(p.failTimeStamps, time.Now())
-	p.lastError = err
-	job.env.registeredPlugins.Store(id, p)
+	status.failTimestamps = append(status.failTimestamps, time.Now())
+	status.Crashed = false
+	job.healthCheckStatuses.Store(id, status)
+
+	if shouldDeactivatePlugin(status.failTimestamps) {
+		status.failTimestamps = []time.Time{}
```
Comment (ali-farooq0) [inline plugin/health_check.go]:
```suggestion
		status.failTimestamps = nil
```

File: plugin/health_check.go [hunk 13]
```diff
@@ -12,105 +12,91 @@ import (
...
+}
 
+// handleHealthCheckFail restarts or deactivates the plugin based on how many times it has failed in a configured amount of time.
+func (job *PluginHealthCheckJob) handleHealthCheckFail(id string, status *PluginHealthCheckStatus) {
 	// Append current failure before checking for deactivate vs restart action
-	p.failTimeStamps = append(p.failTimeStamps, time.Now())
-	p.lastError = err
-	job.env.registeredPlugins.Store(id, p)
+	status.failTimestamps = append(status.failTimestamps, time.Now())
+	status.Crashed = false
```
Comment (ali-farooq0) [inline plugin/health_check.go]: remove

File: plugin/health_check.go [hunk 14]
```diff
@@ -12,105 +12,91 @@ import (
...
+		status.Crashed = true
 
-	if shouldDeactivatePlugin(p) {
-		p.failTimeStamps = []time.Time{}
-		job.env.registeredPlugins.Store(id, p)
+		job.healthCheckStatuses.Store(id, status)
 		mlog.Debug("Deactivating plugin due to multiple crashes", mlog.String("id", id))
 		job.env.Deactivate(id)
-		job.env.SetPluginState(id, model.PluginStateFailedToStayRunning)
+		job.env.setPluginState(id, model.PluginStateFailedToStayRunning)
```
Comment (ali-farooq0) [inline plugin/health_check.go]: this can be removed

File: plugin/health_check.go [hunk 15]
```diff
@@ -12,105 +12,91 @@ import (
...
-	pluginErr := sup.PerformHealthCheck()
-
-	if pluginErr != nil {
-		mlog.Error("Health check failed for plugin", mlog.String("id", id), mlog.Err(pluginErr))
-		job.handleHealthCheckFail(id, pluginErr)
+	status := statusInterface.(*PluginHealthCheckStatus)
+	status.Crashed = false
+	job.healthCheckStatuses.Store(id, status)
+
+	err := job.env.performHealthCheck(id)
```
Comment (ali-farooq0) [inline plugin/health_check.go]: only reset crashed state when plugin is healthy.

File: plugin/health_check.go [hunk 16]
```diff
@@ -12,105 +12,91 @@ import (
...
+	if err != nil {
+		mlog.Error("Health check failed for plugin", mlog.String("id", id), mlog.Err(err))
+		job.handleHealthCheckFail(id, status)
 	}
 }
 
-// handleHealthCheckFail restarts or deactivates the plugin based on how many times it has failed in a configured amount of time.
-func (job *PluginHealthCheckJob) handleHealthCheckFail(id string, err error) {
-	rp, ok := job.env.registeredPlugins.Load(id)
+func (job *PluginHealthCheckJob) hasPluginCrashed(id string) bool {
```
Comment (ali-farooq0) [inline plugin/health_check.go]: `isPluginHealthy`

File: plugin/health_check.go [hunk 17]
```diff
@@ -134,18 +98,24 @@ func (job *PluginHealthCheckJob) Cancel() {
...
+		return false
+	}
+
+	index := len(failedTimestamps) - HEALTH_CHECK_NUM_RESTARTS_LIMIT
+	return time.Since(failedTimestamps[index]) <= HEALTH_CHECK_DEACTIVATION_WINDOW
+}
+
+// removeStaleTimestamps filters out failure timestamps that are before the start of the HEALTH_CHECK_DEACTIVATION_WINDOW
+func removeStaleTimestamps(timestamps []time.Time) []time.Time {
+	result := []time.Time{}
```
Comment (ali-farooq0) [inline plugin/health_check.go]: i think you should just be able to do
```

if len(timestamps) > HEALTH_CHECK_NUM_RESTARTS_LIMIT {
    // Only keeps the last HEALTH_CHECK_NUM_RESTARTS_LIMIT items in timestamps
	timestamps = timestamps[len(timestamps)-HEALTH_CHECK_NUM_RESTARTS_LIMIT:]
}

return timestamps

```

File: plugin/health_check.go [hunk 18]
```diff
@@ -12,113 +12,77 @@ import (
...
+		job.env.setPluginState(id, model.PluginStateFailedToStayRunning)
+		job.failureTimestamps.Delete(id)
 	} else {
 		mlog.Debug("Restarting plugin due to failed health check", mlog.String("id", id))
 		if err := job.env.RestartPlugin(id); err != nil {
 			mlog.Error("Failed to restart plugin", mlog.String("id", id), mlog.Err(err))
 		}
+
+		timestamps = removeStaleTimestamps(timestamps)
+		job.failureTimestamps.Store(id, timestamps)
```
Comment (ali-farooq0) [inline plugin/health_check.go]:
```suggestion
		job.failureTimestamps.Store(id, removeStaleTimestamps(timestamps))
```

File: app/plugin_health_check_test.go [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+	defer tearDown()
+
+	env := th.App.GetPluginsEnvironment()
+	job := env.GetPluginHealthCheckJob()
+	require.NotNil(t, job)
+	bundles := env.Active()
+	require.Equal(t, 1, len(bundles))
+
+	id := bundles[0].Manifest.Id
+	job.CheckPlugin(id)
```
Comment (ali-farooq0) [inline app/plugin_health_check_test.go]:
```suggestion
	// First check
	job.CheckPlugin(id)
```

File: app/plugin_health_check_test.go [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+	bundles = env.Active()
+	require.Equal(t, 1, len(bundles))
+	require.Equal(t, id, bundles[0].Manifest.Id)
+	require.Equal(t, model.PluginStateRunning, env.GetPluginState(id))
+
+	job.CheckPlugin(id)
```
Comment (ali-farooq0) [inline app/plugin_health_check_test.go]:
```suggestion
	// Second check
	job.CheckPlugin(id)
```

File: app/plugin_health_check_test.go [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+	bundles = env.Active()
+	require.Equal(t, 1, len(bundles))
+	require.Equal(t, id, bundles[0].Manifest.Id)
+	require.Equal(t, model.PluginStateRunning, env.GetPluginState(id))
+
+	job.CheckPlugin(id)
```
Comment (ali-farooq0) [inline app/plugin_health_check_test.go]:
```suggestion
	// Should disable plugin
	job.CheckPlugin(id)
```

File: app/plugin_health_check_test.go [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+	bundles = env.Active()
+	require.Equal(t, 0, len(bundles))
+	require.Equal(t, model.PluginStateFailedToStayRunning, env.GetPluginState(id))
+
+	env.Activate(id)
```
Comment (ali-farooq0) [inline app/plugin_health_check_test.go]:
```suggestion
	// Activate manually, should stay activated
	env.Activate(id)
```

File: plugin/health_check.go [hunk 23]
```diff
@@ -12,113 +12,77 @@ import (
...
 	}
 }
 
+// ensureStoredTimestamps returns the stored failure timestamps for a plugin.
+func (job *PluginHealthCheckJob) ensureStoredTimestamps(id string) []time.Time {
+	statusInterface, _ := job.failureTimestamps.LoadOrStore(id, []time.Time{})
```
[the remark this answers is not in the document]
Comment (mickmister) [reply plugin/health_check.go]: We don't want to return `nil` though? Are you suggesting this:
```go
statusInterface, _ := job.failureTimestamps.LoadOrStore(id)
if statusInterface == nil {
    statusInterface = []time.Time{}
}
return statusInterface.([]time.Time)
```
The `sync.Map` allows you to do this in one line and is possibly more performant which is why I chose to use `LoadOrStore`.

File: plugin/health_check.go [hunk 24]
```diff
@@ -12,113 +12,77 @@ import (
...
-// handleHealthCheckFail restarts or deactivates the plugin based on how many times it has failed in a configured amount of time.
-func (job *PluginHealthCheckJob) handleHealthCheckFail(id string, err error) {
-	rp, ok := job.env.registeredPlugins.Load(id)
-	if !ok {
+// CheckPlugin determines the plugin's health status, then handles the error or success case.
+// If the plugin passes the health check, do nothing.
+// If the plugin fails the health check, the function either restarts or deactivates the plugin, based on the quantity and frequency of its failures.
+func (job *PluginHealthCheckJob) CheckPlugin(id string) {
+	err := job.env.performHealthCheck(id)
+	if err == nil {
```
[the remark this answers is not in the document]
Comment (mickmister) [reply plugin/health_check.go]: We've decided to have the convention in plugin repos to have `if err == nil` on its own line which is why I naturally did this. But I will change it :+1:

File: plugin/health_check.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (ali-farooq0) [reply plugin/health_check.go]: @mickmister i think it's safe to return `nil` as we're always using `append` which takes care of the `nil` care automagically ;-)

[the remark this answers is not in the document]
Comment (ali-farooq0) [reply plugin/health_check.go]:
```
statusInterface, _ := job.failureTimestamps.Load(id)
return statusInterface.([]time.Time)
```

File: plugin/health_check.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (mickmister) [reply plugin/health_check.go]: Tried this refactor. We actually need the error for the next line to print the error. The scope of what has access to the vars declared in the if statement is one of the main reasons why we decided to use the more verbose syntax.

[the remark this answers is not in the document]
Comment (ali-farooq0) [reply plugin/health_check.go]: ahh yes, forgot about that :D

File: plugin/health_check.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (mickmister) [reply plugin/health_check.go]: This ended up causing `panic: interface conversion: interface {} is nil, not []time.Time` so I've changed it to guard against non-presence

**Comment by mickmister** [comment]
@lieut-data This PR is ready for review now :+1:

File: app/plugin_health_check_test.go [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+	job.CheckPlugin(id)
+	bundles = env.Active()
+	require.Equal(t, 1, len(bundles))
+	require.Equal(t, id, bundles[0].Manifest.Id)
+	require.Equal(t, model.PluginStateRunning, env.GetPluginState(id))
+
+	// Third health check, plugin should be deactivated by the job
+	hooks, err = env.HooksForPlugin(id)
+	require.Nil(t, err)
+	hooks.MessageWillBePosted(&plugin.Context{}, &model.Post{})
```
Comment (ali-farooq0) [inline app/plugin_health_check_test.go]: i believe these are not needed since the first call to `MessageWillBePosted` would've crashed the plugin, but no harm in keeping it :D

File: plugin/health_check.go [hunk 26]
```diff
@@ -134,18 +103,22 @@ func (job *PluginHealthCheckJob) Cancel() {
...
+	if len(failedTimestamps) < HEALTH_CHECK_NUM_RESTARTS_LIMIT {
+		return false
 	}
-	return false
+
+	index := len(failedTimestamps) - HEALTH_CHECK_NUM_RESTARTS_LIMIT
+	return time.Since(failedTimestamps[index]) <= HEALTH_CHECK_DEACTIVATION_WINDOW
+}
+
+// removeStaleTimestamps only keeps the last HEALTH_CHECK_NUM_RESTARTS_LIMIT items in timestamps
```
Comment (ali-farooq0) [inline plugin/health_check.go]:
```suggestion
// removeStaleTimestamps only keeps the last HEALTH_CHECK_NUM_RESTARTS_LIMIT items in timestamps.
```

File: plugin/health_check.go [hunk 27]
```diff
@@ -12,113 +12,82 @@ import (
...
+		// Store this failure so we can continue to monitor the plugin
+		job.failureTimestamps.Store(id, removeStaleTimestamps(timestamps))
 	}
 }
 
+// getStoredTimestamps returns the stored failure timestamps for a plugin.
+func (job *PluginHealthCheckJob) getStoredTimestamps(id string) []time.Time {
+	timestamps, ok := job.failureTimestamps.Load(id)
+	if !ok {
+		timestamps = []time.Time{}
```
Comment (ali-farooq0) [inline plugin/health_check.go]: ahh sad that the other suggestion didn't work 😞 
```suggestion
		return nil
```

File: app/plugin_health_check_test.go (hunk 25 above)
Comment (mickmister) [reply app/plugin_health_check_test.go]: We need to call it every time we want the plugin to crash though. The plugin gets restarted by the health check job each time we call `CheckPlugin`, so the hook does need to be called for each check here.

Comment (ali-farooq0) [reply app/plugin_health_check_test.go]: ahh forgot about the restart 🤦‍♂. you didn't need it before because `Activate()` was causing the crash, cool makes sense 👌

File: plugin/environment.go [hunk 28]
```diff
@@ -492,5 +496,30 @@ func (env *Environment) SetPrepackagedPlugins(plugins []*PrepackagedPlugin) {
...
+		job := newPluginHealthCheckJob(env)
+		env.pluginHealthCheckJob = job
+		go job.Start()
+	}
+
+	// Config is set to disable. Job exists, kill existing job.
+	if !enable && env.pluginHealthCheckJob != nil {
+		mlog.Debug("Disabling plugin health check job")
+
+		env.pluginHealthCheckJob.Cancel()
```
Comment (lieut-data) [inline plugin/environment.go]: Unrelated to this PR, but I wonder if a simpler model would be to leave the job running all the time, but just have it check if it /should/ do anything on each tick. That would remove all the cleanup code and potential for race conditions if `InitPluginHealthCheckJob` were executed concurrently (i.e. due to quick config updates).

File: plugin/health_check.go [hunk 29]
```diff
@@ -12,113 +12,82 @@ import (
...
-	}
+	cancel            chan struct{}
+	cancelled         chan struct{}
+	cancelOnce        sync.Once
+	env               *Environment
+	failureTimestamps sync.Map
 }
 
 // Start continuously runs health checks on all active plugins, on a timer.
 func (job *PluginHealthCheckJob) Start() {
```
Comment (lieut-data) [inline plugin/health_check.go]: nit: This is no longer a `Start` method but a `Run` method, since it won't return until it's finished.

File: plugin/environment.go [hunk 30]
```diff
@@ -492,5 +508,30 @@ func (env *Environment) SetPrepackagedPlugins(plugins []*PrepackagedPlugin) {
...
+		env.pluginHealthCheckJob = nil
+	}
+}
+
+// GetPluginHealthCheckJob returns the PluginHealthCheckJob owned by the Environment. If the job is not set, the function returns nil.
+func (env *Environment) GetPluginHealthCheckJob() *PluginHealthCheckJob {
```
[the remark this answers is not in the document]
Comment (mickmister) [reply plugin/environment.go]: It is being exported so the test residing in the `app` package can access the method. The test needs to be outside of the `plugin` package so a `Environment` can be instantiated using the testing infrastructure in the `app` package.

File: plugin/environment.go (hunk 28 above)
Comment (mickmister) [reply plugin/environment.go]: Yes I agree. Sounds like something useful for the jobs running in plugins too! Would you like me to create a ticket for this?

**Comment by mickmister** [comment]
@lieut-data Does this need a QA pass before merge?

**Comment by DHaussermann** [comment]
@mickmister Can you please clarify the expected result on this for me? Should I be seeing a change in server logs? I was expecting to see `PluginStateFailedToStayRunning` 

This is what I see now when the health-check fails:
```
{"level":"error","ts":1585595312.3221009,"caller":"plugin/health_check.go:59","msg":"Health check failed for plugin","id":"com.mattermost.demo-plugin","error":"Plugin RPC connection is not responding"}
{"level":"error","ts":1585595312.3221524,"caller":"mlog/log.go:184","msg":"RPC call OnDeactivate to plugin failed.","plugin_id":"com.mattermost.demo-plugin","error":"connection is shut down"}
{"level":"warn","ts":1585595312.3222039,"caller":"plugin/hclog_adapter.go:69","msg":"error closing client during Kill","plugin_id":"com.mattermost.demo-plugin","wrapped_extras":"errconnection is shut down"}
{"level":"warn","ts":1585595312.3222206,"caller":"plugin/hclog_adapter.go:71","msg":"plugin failed to exit gracefully","plugin_id":"com.mattermost.demo-plugin"}
{"level":"error","ts":1585595391.8145926,"caller":"mlog/log.go:184","msg":"Access to route for non-existent plugin","missing_plugin_id":"com.mattermost.demo-plugin","error":"plugin not found: com.mattermost.demo-plugin"}
```

**Comment by mickmister** [comment]
@DHaussermann You need to have the `LogLevel` set to `DEBUG` in order to see the more detailed log prints. Assuming it is set to `DEBUG`, after the first crash you should see `Restarting plugin due to failed health check`. Same for after the second crash. After the third crash, you should see `Deactivating plugin due to multiple crashes`.

The health check is run every 30 seconds, so there will be some time lag between the crashes.

After the third crash, you should also see this in the admin console. It currently requires a refresh to see this, which is not ideal. I'm fairly certain it should show without refreshing so I will look into that.

image
